onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,515 @@
+    // a0_3 a1_3 a2_3 a3_3
+    a3 = vreinterpretq_f32_f64(vzip2q_f64(vreinterpretq_f64_f32(b1), vreinterpretq_f64_f32(b3)));
+
+    return vaddq_f32(vaddq_f32(a0, a1), vaddq_f32(a2, a3));
+}
+
+template <size_t Capacity>
+MLAS_FORCEINLINE std::array<float32x4_t, Capacity / 4>
+LoadDataAndZeroPad(const float* data, size_t count)
+{
+    static_assert(Capacity % 4 == 0, "Capacity must be divisible by 4.");
+
+    count = std::min(count, Capacity);
+
+    std::array<float32x4_t, Capacity / 4> dst{};
+
+    size_t vi = 0;  // vector index
+
+    // handle 4 values at a time
+    while (count > 3) {
+        dst[vi] = vld1q_f32(data);
+
+        vi += 1;
+        data += 4;
+        count -= 4;
+    }
+
+    // handle remaining values
+    if (count > 0) {
+        dst[vi] = vsetq_lane_f32(data[0], dst[vi], 0);
+    }
+
+    if (count > 1) {
+        dst[vi] = vsetq_lane_f32(data[1], dst[vi], 1);
+    }
+
+    if (count > 2) {
+        dst[vi] = vsetq_lane_f32(data[2], dst[vi], 2);
+    }
+
+    return dst;
+}
+
+template <size_t BlkBitWidth, size_t BlkLen, size_t NCols>
+MLAS_FORCEINLINE void
+ComputeDotProducts(
+    const float* ARowPtr,
+    const uint8_t* QuantBDataColPtr,
+    const float* QuantBScaleColPtr,
+    const uint8_t* QuantBZeroPointColPtr,
+    float* SumPtr,
+    size_t CountK,
+    size_t StrideQuantBData,
+    size_t StrideQuantBScale,
+    size_t StrideQuantBZeroPoint,
+    const float* BiasPtr
+)
+{
+    static_assert(NCols == 1 || NCols == 4, "NCols must be 1 or 4");
+
+    const uint8x8_t LowMask = vdup_n_u8(0x0F);
+
+    float32x4_t acc[NCols]{};
+
+    const uint8_t* QuantBData = QuantBDataColPtr;
+    const float* QuantBScale = QuantBScaleColPtr;
+    size_t QuantBZeroPointIdx = 0;  // track half byte increments with this index instead of a pointer
+
+    for (size_t k = 0; k < CountK; k += BlkLen) {
+        const size_t k_blk_len = std::min(CountK - k, BlkLen);
+
+        const uint8_t* b_data[NCols];
+        UnrolledLoop<NCols>(
+            [&](size_t i) { b_data[i] = QuantBData + i * StrideQuantBData; }
+        );
+
+        float scale[NCols];
+        UnrolledLoop<NCols>(
+            [&](size_t i) { scale[i] = QuantBScale[i * StrideQuantBScale]; }
+        );
+
+        float offset[NCols];  // Includes zero point and float conversion offset of 16.
+                              // More details on the float conversion below.
+        if (QuantBZeroPointColPtr != nullptr) {
+            UnrolledLoop<NCols>([&](size_t i) {
+                const uint8_t zp_packed =
+                    QuantBZeroPointColPtr[i * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+                const uint8_t zp = ((QuantBZeroPointIdx & 1) == 1) ? (zp_packed >> 4) : (zp_packed & 0x0F);
+                offset[i] = 16.0f + zp;
+            });
+        } else {
+            UnrolledLoop<NCols>([&](size_t i) {
+                constexpr uint8_t zp = 8;
+                offset[i] = 16.0f + zp;
```

Comment:
>16.0f + zp; [](http://example.com/codeflow?start=28&length=11)

nit: 16.f + 8.f

Reply:
keeping a separate `zp` constant so it has a name but changed it to float